### ISSUE TITLE
[catboost-spark] java.lang.RuntimeException on the big data 

### ISSUE THREAD

**Issue by dspetukhov** [catboost/catboost#1745]
Hi!

I use catboost-spark 0.26 with Spark 2.4.7 and Scala 2.11 in Jupyter with Scala kernel for model training on some big data (with the shape of about 2-3 million rows and ten thousands columns)
During this I encounter the following error:

```
Name: java.lang.RuntimeException
Message: Error while executing master
StackTrace:   at ai.catboost.spark.impl.Helpers$.checkOneFutureAndWaitForOther(Helpers.scala:30)
  at ai.catboost.spark.CatBoostPredictorTrait$class.fit(CatBoostPredictor.scala:208)
  at ai.catboost.spark.CatBoostClassifier.fit(CatBoostClassifier.scala:362)
  ... 42 elided
Caused by: java.util.concurrent.ExecutionException: ai.catboost.CatBoostError: Initial worker wait timeout of 10min 00sec expired
  at java.util.concurrent.FutureTask.report(FutureTask.java:122)
  at java.util.concurrent.FutureTask.get(FutureTask.java:192)
  at ai.catboost.spark.impl.Helpers$.checkOneFutureAndWaitForOther(Helpers.scala:26)
  ... 44 more
Caused by: ai.catboost.CatBoostError: Initial worker wait timeout of 10min 00sec expired
  at ai.catboost.spark.UpdatableWorkersInfo.initWorkers(TrainingDriver.scala:104)
  at ai.catboost.spark.TrainingDriver.run(TrainingDriver.scala:186)
  at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
  at java.util.concurrent.FutureTask.run(FutureTask.java:266)
  at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
  at java.util.concurrent.FutureTask.run(FutureTask.java:266)
```

This error arises in notebook cell output only, no error logs in Jupyter, no error logs in the containers 
I also use the following keys: --master yarn --deploy-mode client 
When I train the model on some small data everything is fine

**Comment by andrey-khropov** [comment]
What's the configuration of your cluster?

CatBoost on Spark does indeed assume by default that training data for CatBoost training worker on each executor can be loaded within 10 minutes timeout.
If loading takes longer CatBoost assumes something went wrong with the executors and aborts the training.
It is not set too high by default to prevent CatBoost wasting too much time waiting for problematic executor hosts.

If the number of workers is small with respect to the data size this timeout can be reached.

This timeout can be increased by setting `WorkerInitializationTimeout` parameter of CatBoostRegressor/CatBoostClassifier.

For example (in Scala):

```scala
    val regressor = new CatBoostRegressor()
      .setWorkerInitializationTimeout(java.time.Duration.ofMinutes(30))
```

You can check task executors logs to look whether training data has been loaded by looking for messages like

```
processPartition: get data providers: start
```

```
processPartition: get data providers: finish
```

If you see only the first message before the application aborts it means that data loading timeout is indeed the cause of the error.

**Comment by dspetukhov** [comment]
Thank you, Andrey
I did some experiments and found that this was exactly the problem. So when I run training as follows:
```
val model = new CatBoostClassifier()
    .setEvalMetric("AUC")
    .setIterations(5)
    .setRandomSeed(0)
    .setWorkerInitializationTimeout(java.time.Duration.ofMinutes(180))
    .setDepth(6).fit(train)
```
no problem arises

However, when I run training as follows:
```
val model = new CatBoostClassifier()
    .setEvalMetric("AUC")
    .setEarlyStoppingRounds(2)
    .setIterations(5)
    .setRandomSeed(0)
    .setUseBestModel(true)
    .setWorkerInitializationTimeout(java.time.Duration.ofMinutes(180))
    .setDepth(6).fit(train, Array[Pool](test))
```
there are too much errors, some of which I will list below
Firstly, in the notebook I can see folowing:
```
Name: java.lang.RuntimeException
Message: Error while executing master
StackTrace:   at ai.catboost.spark.impl.Helpers$.checkOneFutureAndWaitForOther(Helpers.scala:30)
  at ai.catboost.spark.CatBoostPredictorTrait$class.fit(CatBoostPredictor.scala:208)
  at ai.catboost.spark.CatBoostClassifier.fit(CatBoostClassifier.scala:362)
  ... 42 elided
Caused by: java.util.concurrent.ExecutionException: java.io.EOFException
  at java.util.concurrent.FutureTask.report(FutureTask.java:122)
  at java.util.concurrent.FutureTask.get(FutureTask.java:192)
  at ai.catboost.spark.impl.Helpers$.checkOneFutureAndWaitForOther(Helpers.scala:26)
  ... 44 more
Caused by: java.io.EOFException
  at java.io.ObjectInputStream$PeekInputStream.readFully(ObjectInputStream.java:2681)
  at java.io.ObjectInputStream$BlockDataInputStream.readShort(ObjectInputStream.java:3156)
  at java.io.ObjectInputStream.readStreamHeader(ObjectInputStream.java:862)
  at java.io.ObjectInputStream.<init>(ObjectInputStream.java:358)
  at ai.catboost.spark.UpdatableWorkersInfo.acceptAndProcessWorkerInfo(TrainingDriver.scala:74)
  at ai.catboost.spark.UpdatableWorkersInfo.initWorkers(TrainingDriver.scala:93)
  at ai.catboost.spark.TrainingDriver.run(TrainingDriver.scala:186)
  at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
  at java.util.concurrent.FutureTask.run(FutureTask.java:266)
  at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
  at java.util.concurrent.FutureTask.run(FutureTask.java:266)
```
Secondly, in Jupyter log I can see something wrong with the process of data transferring between driver and executors:
```
java.lang.RuntimeException: Error while executing TrainingDriver.waitForListeningPortAndSendWorkerInfo
        at ai.catboost.spark.impl.Helpers$.checkOneFutureAndWaitForOther(Helpers.scala:30)
        at ai.catboost.spark.impl.CatBoostWorker.processPartition(Workers.scala:122)
        at ai.catboost.spark.impl.CatBoostWorkers$$anonfun$run$3.apply(Workers.scala:235)
        at ai.catboost.spark.impl.CatBoostWorkers$$anonfun$run$3.apply(Workers.scala:234)
        at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$28.apply(RDD.scala:980)
        at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$28.apply(RDD.scala:980)
        at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2101)
        at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2101)
        at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
        at org.apache.spark.scheduler.Task.run(Task.scala:123)
        at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)
        at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
        at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.util.concurrent.ExecutionException: java.net.ConnectException: Connection refused (Connection refused)
        at java.util.concurrent.FutureTask.report(FutureTask.java:122)
        at java.util.concurrent.FutureTask.get(FutureTask.java:192)
        at ai.catboost.spark.impl.Helpers$.checkOneFutureAndWaitForOther(Helpers.scala:26)
        ... 15 more
Caused by: java.net.ConnectException: Connection refused (Connection refused)
        at java.net.PlainSocketImpl.socketConnect(Native Method)
        at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
        at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
        at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
        at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
        at java.net.Socket.connect(Socket.java:589)
        at java.net.Socket.connect(Socket.java:538)
        at java.net.Socket.<init>(Socket.java:434)
        at java.net.Socket.<init>(Socket.java:244)
        at ai.catboost.spark.TrainingDriver$.waitForListeningPortAndSendWorkerInfo(TrainingDriver.scala:277)
        at ai.catboost.spark.impl.CatBoostWorker$$anon$1.run(Workers.scala:95)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        ... 3 more
```
Lastly, in the Spark task log I can see following:
```
21/07/13 20:34:14 INFO impl.CatBoostWorker: processPartition: start
21/07/13 20:34:14 INFO impl.CatBoostWorker: processPartition: get data providers: start
21/07/13 20:35:08 INFO impl.CatBoostWorker: processPartition: get data providers: finish
21/07/13 20:35:08 INFO impl.CatBoostWorker: processPartition: CreateTrainingDataForWorker: start
21/07/13 20:35:09 INFO impl.CatBoostWorker: processPartition: CreateTrainingDataForWorker: finish
21/07/13 20:35:09 INFO spark.TrainingDriver: Reserved port 53254 for CatBoost worker
21/07/13 20:35:09 INFO spark.TrainingDriver: wait for CatBoost worker to start listening at port 53254
21/07/13 20:35:09 INFO spark.TrainingDriver: CatBoost worker started listening at port 53254
21/07/13 20:35:09 INFO spark.TrainingDriver: send WorkerInfo to CatBoost training driver
21/07/13 20:35:09 INFO spark.TrainingDriver: WorkerInfo=partitionId=10, partitionSize=40943
21/07/13 20:35:09 INFO spark.TrainingDriver: WorkerInfo has been successfully sent to CatBoost training driver
21/07/13 20:39:32 INFO executor.CoarseGrainedExecutorBackend: Driver commanded a shutdown
21/07/13 20:39:32 ERROR executor.CoarseGrainedExecutorBackend: RECEIVED SIGNAL TERM
21/07/13 20:39:32 WARN util.JavaUtils: Attempt to delete using native Unix OS command failed for path = /disk11/yarn/nm/usercache/appcache/application_1625451552048_6866/blockmgr-ef156f05-07ad-4e2e-9fca-90e733ce81b3. Falling back to Java IO way
java.io.IOException: Failed to delete: /disk11/yarn/nm/usercache/appcache/
```

**Comment by dspetukhov** [comment]
Moreover, I reduced the number of features to 10, so the training dataset was a shape of 600k x 10 and validation dataset - 200k x 10
Unfortunately, during training with early stopping I have faced an error `org.apache.spark.shuffle.MetadataFetchFailedException: Missing an output location for shuffle 12`
In the Spark task log I have found the following error during `foreachPartition at Workers.scala:233`:

```
21/07/20 15:52:39 INFO codegen.CodeGenerator: Code generated in 23.389976 ms
21/07/20 15:52:39 INFO impl.CatBoostWorker: processPartition: start
21/07/20 15:52:39 INFO impl.CatBoostWorker: processPartition: get data providers: start
21/07/20 15:52:39 INFO impl.CatBoostWorker: processPartition: get data providers: finish
21/07/20 15:52:39 INFO impl.CatBoostWorker: processPartition: CreateTrainingDataForWorker: start
21/07/20 15:52:39 INFO impl.CatBoostWorker: processPartition: CreateTrainingDataForWorker: finish
21/07/20 15:52:39 INFO spark.TrainingDriver: Reserved port 17075 for CatBoost worker
21/07/20 15:52:39 INFO spark.TrainingDriver: wait for CatBoost worker to start listening at port 17075
uncaught exception:
    address -> 0x107266410
    what() -> "(Bad file descriptor) util/network/socket.h:76: setsockopt() failed for reuse addr"
    type -> TSystemError
```

and this is similar to https://github.com/catboost/catboost/issues/1765
and also I see the following errors in Jupyter:
```
[CatBoost Master] ERROR: 2021-07-21 13:15:08.661 +0300 par_network.cpp:201 timeout(10.000000s seconds) exceed, info:reqId: a66840d7-8d7dc1e3-bec592a6-c97f9b6e tcp2://10.218.64.14:41119/matrixnet init@41025 retries rest: 40
[CatBoost Master] ERROR: 2021-07-21 13:15:10.654 +0300 par_network.cpp:201 timeout(10.000000s seconds) exceed, info:reqId: a9c8ea7c-e62d062d-e2bb1edc-27eddd94 tcp2://10.218.64.14:41119/matrixnet _ping_ retries rest: 40
...
[CatBoost Master] FAIL (2021-07-21T13:17:05.051447+0300): got unexpected network error, no retries rest
[CatBoost Master]   library/cpp/par/par_network.cpp:187
[CatBoost Master]   MultiClientThreadLoopFunction() failed
[CatBoost Master] ??+0 (0x7F8BA58CE7E9)
[CatBoost Master] ??+0 (0x7F8BA58C7DBA)
[CatBoost Master] ??+0 (0x7F8BA64203ED)
[CatBoost Master] ??+0 (0x7F8BA598A0F6)
[CatBoost Master] ??+0 (0x7F8BA598A3C7)
[CatBoost Master] ??+0 (0x7F8BA58CE94B)
[CatBoost Master] ??+0 (0x7F8CBD484184)
[CatBoost Master] clone+109 (0x7F8CBDBB503D)
```

Not sure that this is intrinsically catboost-spark error

**Comment by dspetukhov** [comment]
Ok, I guess that the issue can be closed as the initial question was answered
The problems described below are related to previous version of the library and may have been resolved in 0.26.1
